expression/builtin.go

```diff
@@ -258,6 +262,34 @@ func newBaseBuiltinFuncWithTp(ctx sessionctx.Context, funcName string, args []Ex
 	return bf, nil
 }
 
+func generateRetFlag(res Expression, evalType types.EvalType, args []Expression) {
+	if len(args) < 2 {
+		return
+	}
+	var retType = res.GetType()
+	lhs, rhs := args[0], args[1]
+	if fun, ok := res.(*ScalarFunction); (ok && fun.FuncName.L == ast.Cast) || res.GetType().Tp == mysql.TypeBit {
+		if evalType == types.ETInt || evalType == types.ETReal {
+			lfg, rfg := args[0].GetType().Flag, args[1].GetType().Flag
+			// set signedFlag to 0 when bit convert explicitly or implicitly to int or real
+			if mysql.TypeBit == lhs.GetType().Tp || mysql.TypeBit == rhs.GetType().Tp {
+				retType.Flag |= mysql.UnsignedFlag
+				if mysql.TypeBit == rhs.GetType().Tp {
+					retType.Flag &= ^(mysql.UnsignedFlag ^ lfg)
+				} else {
+					retType.Flag &= ^(mysql.UnsignedFlag ^ rfg)
+				}
+			}
+		} else if evalType == types.ETDecimal && (mysql.TypeEnum == lhs.GetType().Tp || mysql.TypeEnum == rhs.GetType().Tp) {
+			// set Flen and Decimal as unspecified length(-1) when the enum type is compare with decimal type,avoid loss of data,
+			// like the enum'z'(26), wher Flen=2,Decimal=2 the enum=26.00 but when Flen= 1,Decimal=2 the enum=9
```

Comment:
I think the logic is wrong..  When comparing enum and decimal, we should cast both enum and decimal to double....
In mysql 8.0, 
```
MySQL [test]> desc format=tree select * from t where t.b > t.c;
+------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                |
+------------------------------------------------------------------------------------------------------------------------+
| -> Filter: (cast(t.b as double) > cast(t.c as double))  (cost=0.35 rows=1)
    -> Table scan on t  (cost=0.35 rows=1)
 |
+------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.004 sec)

```
In tidb, enum will convert to decimal.
```
[tidb]> desc select * from t where t.b > t.c;
+-------------------------+---------+-----------+---------------+--------------------------------+
| id                      | estRows | task      | access object | operator info                  |
+-------------------------+---------+-----------+---------------+--------------------------------+
| Selection_5             | 0.80    | root      |               | gt(cast(test.t.b), test.t.c)   |
| └─TableReader_7         | 1.00    | root      |               | data:TableFullScan_6           |
|   └─TableFullScan_6     | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-------------------------+---------+-----------+---------------+--------------------------------+
3 rows in set (0.000 sec)
```

Reply:
done,I don't know why convert enum,decimal to double but not convert enum to decimal

Reply:
done,I don't know why convert enum,decimal to double but not convert enum to decimal